rasa/utils/train_utils.py

```diff
@@ -20,125 +21,225 @@
 
 # avoid warning println on contrib import - remove for tf 2
 tf.contrib._warning = None
+
 logger = logging.getLogger(__name__)
 
-# namedtuple for all tf session related data
-SessionData = namedtuple("SessionData", ("X", "Y", "label_ids"))
+
+# type for all tf session related data
+SessionDataType = Dict[Text, List[np.ndarray]]
 
 
 def load_tf_config(config: Dict[Text, Any]) -> Optional[tf.compat.v1.ConfigProto]:
     """Prepare `tf.compat.v1.ConfigProto` for training"""
+
     if config.get("tf_config") is not None:
         return tf.compat.v1.ConfigProto(**config.pop("tf_config"))
     else:
         return None
 
 
+def create_label_ids(label_ids: "np.ndarray") -> "np.ndarray":
+    """Convert various size label_ids into single dim array.
+
+    for multi-label y, map each distinct row to a string repr
+    using join because str(row) uses an ellipsis if len(row) > 1000.
+    Idea taken from sklearn's stratify split.
+    """
+
+    if label_ids.ndim == 1:
+        return label_ids
+
+    if label_ids.ndim == 2 and label_ids.shape[-1] == 1:
+        return label_ids[:, 0]
+
+    if label_ids.ndim == 2:
+        return np.array([" ".join(row.astype("str")) for row in label_ids])
+
+    if label_ids.ndim == 3 and label_ids.shape[-1] == 1:
+        return np.array([" ".join(row.astype("str")) for row in label_ids[:, :, 0]])
+
+    raise ValueError("Unsupported label_ids dimensions")
+
+
 # noinspection PyPep8Naming
 def train_val_split(
-    session_data: "SessionData", evaluate_on_num_examples: int, random_seed: int
-) -> Tuple["SessionData", "SessionData"]:
+    session_data: SessionDataType,
+    evaluate_on_num_examples: int,
+    random_seed: int,
+    label_key: Text,
+) -> Tuple[SessionDataType, SessionDataType]:
     """Create random hold out validation set using stratified split."""
 
-    label_counts = dict(
-        zip(*np.unique(session_data.label_ids, return_counts=True, axis=0))
+    if label_key not in session_data or len(session_data[label_key]) > 1:
+        raise ValueError(f"Key '{label_key}' not in SessionData.")
+
+    label_ids = create_label_ids(session_data[label_key][0])
+
+    label_counts = dict(zip(*np.unique(label_ids, return_counts=True, axis=0)))
+
+    check_train_test_sizes(evaluate_on_num_examples, label_counts, session_data)
+
+    counts = np.array([label_counts[label] for label in label_ids])
+
+    multi_values = []
+    [
+        multi_values.append(v[counts > 1])
+        for values in session_data.values()
+        for v in values
+    ]
```

Comment:
this is a very odd way to write this imo: either use a normal for & append or just a list comprehension,
```python
 multi_values = [
        v[counts > 1]
        for values in session_data.values()
        for v in values
    ]
```